### PR TITLE
mrc-1494: support for interpolation

### PR DESCRIPTION
**Pull request by richfitz**
This PR adds support for constant and linear interpolation (~spline interpolation requires a bit more bookkeeping, and either an external spline library or something to do a matrix inverse and some tedious book-keeping~).  The interface is the same as for odin (and this PR copies over the tests), and the implementation is a direct port of that in [cinterpolate](https://github.com/mrc-ide/cinterpolate) and includes things like a caching binary search for looking up the appropriate point.

### CODE REVIEW

**Review by giovannic — APPROVED** [review]

File: R/generate_js.R [hunk 1]
```diff
@@ -17,9 +17,15 @@ generate_js <- function(ir, options) {
...
+  fns <- unlist(lapply(dat$equations, function(x) x$depends$functions),
+                FALSE, FALSE)
+  uses_sum <- "sum" %in% fns
+
   ## This is all we need to dump out
   list(code = generate_js_generator(core, dat),
-       name = dat$config$base)
+       name = dat$config$base,
+       include = c(interpolate.js = dat$features$has_interpolate,
+                   support_sum.js = uses_sum))
```
Comment (giovannic) [inline R/generate_js.R:28]: these changes break the [less basic model](https://github.com/mrc-ide/squire/blob/9bd90684db35267b7f408ca90a77990af9549701/inst/odin/less_basic_model_for_js.R)

```
ReferenceError: odinSum2 is not defined
    at less_basic_model_for_js.require.odin.less_basic_model_for_js.less_basic_model_for_js.rhs (/opt/build/squire.js:1077:32)
    at Dopri5.rhs (/opt/build/squire.js:758:13)
    at Dopri5.reset (/opt/build/squire.js:478:14)
    at Dopri.initialise (/opt/build/squire.js:146:23)
    at Object.integrate (/opt/build/squire.js:116:12)
    at Object.integrate (/opt/build/squire.js:559:24)
    at integrateOdin (/opt/build/squire.js:771:21)
    at less_basic_model_for_js.require.odin.less_basic_model_for_js.less_basic_model_for_js.run (/opt/build/squire.js:1119:12)
    at runModel (/opt/build/squire.js:1189:14)
    at Object.<anonymous> (/opt/e2e/test.js:3:15)
```

**Review by giovannic — CHANGES_REQUESTED** [review]
Could you check out the potential regression I've commented?

**Review by giovannic — APPROVED** [review]
Works!

File: inst/interpolate.js [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+    };
+
+    // Optimisation for the constant interpolation - this should be
+    // moved into the C version of this too
+    if (type === "constant") {
+        var ylinear = y;
+        y = [];
+        for (var i = 0; i < n; ++i) {
+            var yi = [];
+            for (j = 0; j < ny; ++j) {
```
Comment (giovannic) [inline inst/interpolate.js:58]: I missed this, but there should be a `var j` here

**Review by giovannic — COMMENTED** [review]
small bug here

File: inst/interpolate.js (hunk 2 above)
Comment (richfitz) [reply inst/interpolate.js:58]: thanks - Ill add that into #7 in a minute :)

Comment (richfitz) [reply inst/interpolate.js:58]: OK, that's fixed on the other PR now

**Pull request merged**